### PR TITLE
Implement incremental reparsing for remaining braced blocks

### LINKED ISSUE #66: Extend incremental reparsing to all braced blocks
The incremental reparsing model of libsyntax2 is curly-block based.

* in the parser, we make sure that, for any pair of `{` and `}`, there's a syntax node for which curlys are the first and the last child.
* this gives us corollary: if edit is inside a matching pair of braces, and the pair remains matching after the edit is applied, the effects of the edit are contained to a single block. 
* this gives a rise to incremental re parsing algorithm: find closest curly-block, verify its structure, reparse.

Currently only expr blocks and struct blocks participate in this algorithm, but it should be extended to any kind of braced block. Here's the relevant code:

https://github.com/matklad/libsyntax2/blob/4f647096665b2ca3725ba1f7415a21fbc46044bb/crates/libsyntax2/src/lib.rs#L145-L147

it should be extended to all braced blocks. The relevant tests are here:

https://github.com/matklad/libsyntax2/blob/4f647096665b2ca3725ba1f7415a21fbc46044bb/crates/libsyntax2/tests/test/main.rs#L42

It would be nice to use `proptest` or fuzzing to check incremental re parsing! (cc @killercup  on the last one). Here's the top-level repasing API:

https://github.com/matklad/libsyntax2/blob/4f647096665b2ca3725ba1f7415a21fbc46044bb/crates/libsyntax2/src/lib.rs#L81

### PR DESCRIPTION
**Pull request by darksv**
Fixes #66

### CODE REVIEW

File: crates/libsyntax2/src/lib.rs [hunk 1]
```diff
@@ -144,6 +144,21 @@ fn find_reparsable_node(node: SyntaxNodeRef, range: TextRange) -> Option<(Syntax
         let res = match node.kind() {
             BLOCK => grammar::block,
             NAMED_FIELD_DEF_LIST => grammar::named_field_def_list,
+            NAMED_FIELD_LIST => grammar::named_field_list,
+            ENUM_VARIANT_LIST => grammar::enum_variant_list,
+            MATCH_ARM_LIST => grammar::match_arm_list,
+            USE_TREE_LIST => grammar::use_tree_list,
+            EXTERN_ITEM_LIST => grammar::extern_item_list,
+            TOKEN_TREE => grammar::token_tree,
```
Comment (matklad) [inline crates/libsyntax2/src/lib.rs]: I think this is slightly wrong. Token tree might use `() or `[]`, and not `{}`. We maintain `{}` structure carefully, but that's not true for `()` and `[]`.  I think we should make an additional check here that first child is `{`.

The test for it probbaly would be like

```
fn foo() {
    vec![<|><|>]
}
```

current implementation I think will try to reparse `[]` and will faile the `is_balanced` check. The correct impl should reparse the fn body.

**Comment by darksv** [comment]
OK, I will look into it later.

**Comment by CAD97** [comment]
CI was green...
https://travis-ci.org/matklad/libsyntax2/builds/426894269
https://ci.appveyor.com/project/matklad/libsyntax2/build/1.0.466

I forgot if I got permission to do this but I'm going to try
bors retry

**Comment by CAD97** [comment]
OK, no idea why bors is timing out.

AppVeyor: https://ci.appveyor.com/project/matklad/libsyntax2/build/1.0.467  
 2 hours ago in 6 min 19 sec 

Travis: https://travis-ci.org/matklad/libsyntax2/builds/426955378  
 Ran for 5 min 39 sec
about 2 hours ago 

(bors[bot] commented an hour ago)

**Comment by matklad** [comment]
bors retry

**Pull request merged**